### ISSUE TITLE
Can't extract

### ISSUE THREAD

**Issue by olibclarke** [tbepler/topaz#12]
Hi,

I ran the tutorial without issue. When I run on my data, the training completes successfully and generates models, but when I run the extraction job, nothing happens - by which I mean topaz uses the GPU, sits there for some time, but writes nothing to the log and no files are generated. Am I doing something obviously dense? And is there a way to make the extraction job more verbose (I can't see a -v flag in the help)? This is the command I used:

topaz extract -r 7 -x 16 -m saved_models/model_epoch10.sav  -o topaz/predicted_picks_upsampled.txt processed/mics/*mrc >& log &

Cheers
Oli

**Comment by olibclarke** [comment]
I notice that the last line of epoch10 in model_training.txt looks weird. And if I extract using epoch9.sav it works. So maybe something went wrong at the end of training?

<img width="1027" alt="image" src="https://user-images.githubusercontent.com/19766818/55237761-48bc6400-5209-11e9-95d3-a8b51cd1acaa.png">

**Comment by olibclarke** [comment]
Hmm never mind I think the training must have failed somehow, the picks look kind of random even with high scores. Does topaz use the same coordinate conventions as crYOLO? I used crYOLO boxfiles as the training data

**Comment by tbepler** [comment]
Hmmm, that looks bad. I don't know what conventions crYOLO uses, but here are a few things to check. This could be because the training particle coordinates are mismatched with the preprocessed micrographs.

1. Make sure that the training coordinates are scaled to match the preprocessed micrographs. This can be done using the 'convert' command (e.g. if you downsampled the micrographs 8x, the coordinates need to be scaled the same amount, -x 8 to convert).

2. Make sure coordinates mark particle centers. If you were to take the first two columns of a box file, for example, that would mark the upper left corner of the box and would not work for training.

3. Particle coordinates need to have the origin in the upper left of each micrograph (0,0 coordinate). This is the usual convention for MRC files, but some other programs/image formats (typically tiff images) place the origin in the lower left corner instead. This will cause the labeled coordinates to be wrong for training and can be fixed using 'convert' with the --invert-y flag. You'll need to pass the micrographs as well so that the height of each micrograph can be used to flip the y-axis coordinates.

4. Make sure you are using a reasonable score threshold. If you plot everything from 'extract' it will contain a lot of false positives, because extract goes to a very low scores. I recommend thresholding at a score of 0 (this can be done with 'convert' flag -t 0). You say they look random even with high scores, so this probably doesn't apply here.

Also, how large are your particles in your preprocessed micrographs in pixels? I recommend setting the extract radius (-r X) to the radius of the particle in the processed micrographs, but that's unlikely to be the cause of this.

**Comment by olibclarke** [comment]
Hi Tristan, I believe they are just EMAN format box files. Here is a sample (before preprocessing). 

I took these box files and converted to star with downscaling: `topaz convert -s 16 --from box --to star 19j*box -o particles_downscaled.star` and then used that as input for training. 

Does convert in this instance adjust so that the particle center is correct in the star file? Re 4, I was using a threshold of 3 in the attached image (~top 5% of picks). 

My particles are very asymmetric (elongated) - ~6 pixels radius in one orientation and ~20 in the long axis, and I extracted with a radius of 7 pixels.

I was not completely clear for training, whether I should supply all the images, or just the ones labeled for training. In this case I completely picked several micrographs and supplied all to training. I'm currently running a job with all micrographs incompletely picked (setting aside one fifth for training).

Cheers
Oli

<img width="227" alt="image" src="https://user-images.githubusercontent.com/19766818/55250519-b7f38180-5224-11e9-8e13-1a69b815ef7d.png">

**Comment by tbepler** [comment]
Convert will adjust the coordinates from box files to be the center of the box, so that shouldn't be the issue. Just as a sanity check, can you verify by eye that the training coordinates are correct on the preprocessed micrographs?

What parameters are you using for the training step?

**Comment by olibclarke** [comment]
`topaz train -n 300 --num-workers=40 -d 0 -k 5 --train-images processed/mics/ --train-targets picks/particles_downscaled.star --save-prefix=saved_models3/model -o saved_models3/model_training.txt >& log&`

The training coordinates look fine when I look at the preprocessed box files on the preprocessed mics - see example: 
<img width="462" alt="image" src="https://user-images.githubusercontent.com/19766818/55252525-84ffbc80-5229-11e9-853d-b34cf16032cb.png">

**Comment by tbepler** [comment]
Definitely not an issue with the labels, then.

Based on your parameters, I suspect that the problem is due to have -n set too large. -n is the expected number of particles per micrograph and the model is trained to detect approximately this number of particles on average per micrograph. This is sort of the core of the positive-unlabeled learning framework. Model training is robust to misspecification of this parameter within a reasonable range, but setting it too large will cause the model to pick a lot of false positives (basically, if there are only 100 particles per micrograph, but you set n=500, then the model is trained to find 400 more things per micrograph and call them positives).

Based on your above micrograph, I would estimate n ≈ 100 (or maybe slightly less). You have n=300, so this is likely the source of the problem. I suggest training again with -n 100 (or a better estimate of n if this micrograph isn't representative of the particle density).

**Comment by olibclarke** [comment]
Hmmm - I set this number based on the maximum number of particles per mic - that is I do have a significant fraction of micrographs with 200-350 true particles (although many with 50-100 also). In this case what is the best approach?

**Comment by olibclarke** [comment]
To be clear, does -n limit the number of particles that will be found on mics that happen to have a higher particle concentration?

**Comment by tbepler** [comment]
No, it doesn't. It is the average number of particles per micrograph across all micrographs. Individually, micrographs can have more/less particles.

**Comment by olibclarke** [comment]
Got it. I will try training with n=100 then

**Comment by olibclarke** [comment]
(the average number of particles per mic calculated from crYOLO picking is indeed ~100)

**Comment by olibclarke** [comment]
(regarding the original error though - it would be good if `topaz extract` reported an error if extraction fails - e.g. if the input model is corrupt or somesuch - rather than silently failing)

**Comment by tbepler** [comment]
Well, we haven't figured out what actually went wrong there. If you run 'topaz extract' with your 10th epoch model without the stdout/stderr redirect, do you get any output at all? Extract will write the coordinates and scores as it finishes each micrograph.

It's possible that you are seeing no output because of stdout/stderr buffering. How long did you let extract run for?

**Comment by olibclarke** [comment]
I let it run until it died and there was zero output to the log or files written out. Running without redirect now

**Comment by tbepler** [comment]
Sorry, you also need to not use the -o flag.

**Comment by tbepler** [comment]
Until it died? Did it give some sort of error eventually?

**Comment by olibclarke** [comment]
No, it just died silently, as in the process was no longer running or using the GPU

**Comment by olibclarke** [comment]
How long should it take to generate output? I will leave it running, but after several minutes there is no output:
<img width="1048" alt="image" src="https://user-images.githubusercontent.com/19766818/55254873-6ef4fa80-522f-11e9-8f05-fbafc93326ec.png">

**Comment by tbepler** [comment]
Interesting, that makes it seem like it completed succesfully. What happens if you run extract on a single micrograph? Does this reproduce?

**Comment by olibclarke** [comment]
Whoops spoke too soon, now it is generating output to the prompt - looks like what I would expect to see in the txt file - mic name, coords and score. But when I ran it with -o and the redirect it generated nothing and just died

**Comment by tbepler** [comment]
My guess is it was some sort of buffering issue or something weird with your file system. It sounds like extract is working correctly. On some clusters, I've had issues before where the root node sees the file is empty but the worker node has written a bunch of stuff and somehow the file has sync'd across the whole system. Could be something like that?

**Comment by olibclarke** [comment]
Sounds reasonable - when I run it on one mic it works fine.

**Comment by olibclarke** [comment]
Even training with n=100, the picks still look garbage. I am extracting with a radius of 6, as that is the radius of the "compact" view of the particle. Not sure what I am doing wrong - this should be an easy one to pick so something must be going fundamentally wrong. 

the second picture shows coords after extracting with a radius of 20 (corresponding to the long axis of the particle) - still not great. Suggestions appreciated, thanks for your help so far!


<img width="466" alt="image" src="https://user-images.githubusercontent.com/19766818/55259050-f8f69080-523a-11e9-8d53-93d9d1f9048c.png">
<img width="462" alt="image" src="https://user-images.githubusercontent.com/19766818/55259291-a9649480-523b-11e9-887f-5987e5090338.png">

**Comment by tbepler** [comment]
What threshold is this at? Can you also share the score histogram (see complete walkthrough tutorial for an example)?

**Comment by tbepler** [comment]
How large is your dataset? Can you share it with me to see if I can reproduce this on my end?

**Comment by olibclarke** [comment]
This is taking the top 500k picks (I know there are ~500k good particles in the final dataset from previous processing). I would be happy to share the dataset (5000 mics) with you although I suspect something stupid must be going wrong. I tried making a set of training data with inverted Y to rule out that possibility, but I got this error when I ran convert : 
<img width="1283" alt="image" src="https://user-images.githubusercontent.com/19766818/55260400-b3d45d80-523e-11e9-80d9-8bef022a199a.png">

**Comment by tbepler** [comment]
Well that's a bug. I'll put that one on the list too.

**Comment by tbepler** [comment]
Could you just share one micrograph with your training particles (box file and/or star file)? Just so I can take a look as a sanity check.

**Comment by tbepler** [comment]
You can just email it to me.

**Comment by olibclarke** [comment]
Here is the score histogram for extraction with a radius of 20:
<img width="904" alt="image" src="https://user-images.githubusercontent.com/19766818/55260630-41b04880-523f-11e9-8cdd-642c31b57d11.png">

**Comment by tbepler** [comment]
That looks pretty decent. Can you set the threshold to ~ -2 and see what the particles look like?

**Comment by tbepler** [comment]
Well, there aren't even 500k particles extracted. Let alone at a good threshold so you're taking all particles at 500k. That said, the picks look so bad on your micrograph that I think something is still being misspecified at the training step.

**Comment by olibclarke** [comment]
it's not *terrible*, but it is still missing plenty of particles that it ought to catch (and which are picked by relion and crYOLO), so I suspect something must be going wrong - I'll email you a mic and box file 
<img width="461" alt="image" src="https://user-images.githubusercontent.com/19766818/55261118-8092ce00-5240-11e9-871e-5ef33a838fbe.png">

**Comment by tbepler** [comment]
Yep, that looks way better. You could try going even lower. Maybe -5

**Comment by tbepler** [comment]
If you try adjusting the threshold for the radius 6 extracted particles, it may do better on the closely packed particles.

**Comment by olibclarke** [comment]
OK i'll give that a go - just sent you a mic and box for sanity check

**Comment by tbepler** [comment]
Ok, based on looking at your micrograph and coordinates, the input data seems fine. I have two hypothesis for what is causing this:

1) your threshold was set too low and extraction with radius 20 loses the closely packed particles. 

This should be fixable by trying your radius 6 extraction and setting the score threshold to a reasonable number.

2) because many instances of the particle are closely packed (i.e. you have single particles and then instances of 2 or more particles very close together) and the training data tends to only contain single particle instances, the model is not detecting closely packed particles. This would explain why you are getting less particles than you expect. Your micrograph above suggests that this might be going on.

If you don't want this behaviour (though this seems to usually be a desirable property), then you could try downsampling the micrographs less (8x maybe even only 4x downsampling). This gives the model less surrounding context and also gives better spatial resolution for separating nearby particles.

Personally, my guess is that radius 6 extraction + using a reasonable threshold will get the job done. If not, reducing the downsampling amount may improve performance (I've seen this improve results for some other datasets at least).

**Comment by olibclarke** [comment]
Ok thanks Tristan - this has been very helpful, thank you. I will give that a go and see how we go. I am skeptical that just changing the threshold will help for radius 6 because many particles do not have a pick in the center of the particle from the raw data... maybe I will try training with less downsampling. Thanks!

**Comment by tbepler** [comment]
No problem! Adjusting the threshold for the radius 6 extraction is easy to do and doesn't require retraining, so it may be worth looking at first. The histogram is a great way to get an idea of what should be reasonable thresholds (look for the bimodality, high score mode = particles, low score mode = junk). Otherwise, hopefully less downsampling will help!

**Comment by olibclarke** [comment]
Hi Tristan,

Here is the histogram with a radius of 6 pixels. So looking at this I'd probably use a threshold of -3. But looking at the help for `topaz extract`, it says the default threshold for `-t` is 0.5, and if I explicitly set a threshold of 0.5, there are still particles with negative scores in the output txt. So is there another threshold parameter, am I missing something obvious?

Cheers
Oli
<img width="662" alt="image" src="https://user-images.githubusercontent.com/19766818/55278094-6c051300-52de-11e9-97a0-0312088f0f1b.png">

**Comment by olibclarke** [comment]
Evidently I am doing something wrong because when I try to extract with a threshold of -3 I get this error. I can just manually grab the lines with scores greater than -3 using awk, maybe that's a better way to go 
<img width="1044" alt="image" src="https://user-images.githubusercontent.com/19766818/55278501-85f52480-52e3-11e9-9b03-f072f751c810.png">

**Comment by tbepler** [comment]
Are you using a different model than yesterday? Why are the histogram scores so different than for the 20 radius extraction?

Regarding the threshold: I don't suggest using the threshold parameter in the extract command. It should probably be renamed but it does not threshold by score. Instead, it is a quantile-based threshold designed strictly to improve the performance of extract by not including coordinates for things that are obvious junk (by default, this is done by discarding all scores below the 50th percentile).

Instead, use the convert command with the -t argument. This is also faster because the micrographs don't have to be rescored etc. It just filters the coordinates table.

**Comment by olibclarke** [comment]
I don't know why they're different - I wasn't sure that they would be expected to be the same? This was just taking the last 100k lines of the file so maybe not representative? In any case using a radius of  6 and a threshold of -3 it definitely looks better! 
<img width="462" alt="image" src="https://user-images.githubusercontent.com/19766818/55279809-361e5980-52f3-11e9-95ae-de252c9cb309.png">

**Comment by olibclarke** [comment]
Not due to lines not being representative - here is the histogram from 100k random lines: 
<img width="855" alt="image" src="https://user-images.githubusercontent.com/19766818/55279825-839ac680-52f3-11e9-84da-7e0e577f3d0f.png">

**Comment by tbepler** [comment]
Ah, I didn't realize this was a subsample. The number of coordinates per bin was so different it seemed a bit weird but that explains it.

Great, though! Those picks look good!

**Comment by olibclarke** [comment]
Yes pretty good, although crYOLO still does better for this dataset (in terms of resolution from the raw particle set) - using less downsampling may help, so I will give that a go!

**Comment by tbepler** [comment]
You may also be able to get better results just with tuning the threshold a bit more. It seems like there some room to take the threshold a bit lower.

Anyway, unless you have more questions regarding train/extract, I'm going to close this issue. It seems like everything is working and it's just a matter of tuning the parameters a bit.

**Comment by olibclarke** [comment]
By all means - thank you for all your help!
